### DISCUSSION TITLE
GSO on-going support （quic-go < 0.38)

### DISCUSSION THREAD

**Discussion by piyoki** [juicity/juicity#42] [General]
## Background

GSO (General Segmentation Offload) is a technology that offloads IP segmentation to the network card, which can greatly reduce the CPU usage and improve the performance of juicity.

> **Warning**
> quic-go currently has limited support for GSO, causing programs based on quic-go to perform poorly.
>
> Fortunately, quic-go currently allows users to manually set GSO environment variables to enable GSO, using hardware offloading for better performance.

GSO can be enabled separately from the `client` side and/or the `server` side, unless some hardware does not support it.

## Implementation

> **Note**
> Since `v0.1.1`, you can enable GSO support by setting an environment variable as follows:

### Windows

```shell
set QUIC_GO_ENABLE_GSO=true
./juicity run -c config.json
```

### Linux/macOS

```shell
export QUIC_GO_ENABLE_GSO=true
./juicity run -c config.json
```

### Systemd Configuration (Linux ONLY)

```conf
[Unit]
Description=juicity-server Service
Documentation=https://github.com/juicity/juicity
After=network.target nss-lookup.target

[Service]
Type=simple
User=root
Environment=QUIC_GO_ENABLE_GSO=true
ExecStart=/usr/bin/juicity-server run -c /etc/juicity/server.json --disable-timestamp
Restart=on-failure
LimitNPROC=512
LimitNOFILE=infinity

[Install]
WantedBy=multi-user.target
```

## Concerns

GSO needs network card feature support. On `Linux`, check the support of GSO with the following command:

```shell
ethtool -k <interface name>
```

But note that showing ON does not mean full support. This is subject to the actual situation.

If the network card does not support GSO and juicity is forced to use GSO through an environment variable, you will get an `error` like:

```shell
invalid argument
```

Some devices disable GSO by default, and you can enable it with the following command. If it doesn't work, you can only give up GSO support on this side.

```shell
ethtool -K <interface name> gso on
```

## Future work

The quic-go community is actively following the support of GSO, see <https://github.com/quic-go/quic-go/issues/3911> for details. Once it is resolved, users will not need to explicitly set the above environment variables.

**Comment by mzz2017** [comment]
This has been fixed in https://github.com/quic-go/quic-go/releases/tag/v0.38.0
